src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -395,6 +399,41 @@ private ReadOnlySpan<byte> GetFrameData(ReadOnlySpan<byte> frameData, bool hasPa
             return frameData;
         }
 
+        /// <summary>
+        /// Parses an ALTSVC frame, defined by RFC 7838 Section 4.
+        /// </summary>
+        /// <remarks>
+        /// The RFC states that any parse errors should result in ignoring the frame.
+        /// </remarks>
+        private void ProcessAltSvcFrame(FrameHeader frameHeader)
+        {
+            if (NetEventSource.IsEnabled) Trace($"{frameHeader}");
+            Debug.Assert(frameHeader.Type == FrameType.AltSvc);
+
+            ReadOnlySpan<byte> span = _incomingBuffer.ActiveSpan.Slice(0, frameHeader.Length);
+
+            if (BinaryPrimitives.TryReadUInt16BigEndian(span, out ushort originLength))
+            {
+                span = span.Slice(2);
+
+                // Check that this ALTSVC frame is valid for our pool's origin. ALTSVC frames can come in one of two ways:
+                //  - On stream 0, the origin will be specified. HTTP/2 can service multiple origins per connection, and so the server can report origins other than what our pool is using.
+                //  - Otherwise, the origin is implicitly defined by the request stream and must be of length 0.
+
+                if ((frameHeader.StreamId != 0 && originLength == 0) || (frameHeader.StreamId == 0 && span.Length >= originLength && span.Slice(originLength).SequenceEqual(_pool.Http2AltSvcOriginUri)))
+                {
+                    span = span.Slice(originLength);
+
+                    // The span now contains a string with the same format as Alt-Svc headers.
+
+                    string altSvcHeaderValue = Encoding.ASCII.GetString(span);
+                    _pool.HandleAltSvc(new[] { altSvcHeaderValue }, null);
```

Comment:
Is one value common?  It's unfortunate we're having to allocate an extra array in this case just to pass it in.  Could the target take a span instead of an enumerable?

Reply:
> Is one value common? It's unfortunate we're having to allocate an extra array in this case just to pass it in.

Yes. I've left a TODO to optimize.

I don't think many H2 servers are sending out ALTSVC frames right now for anything other than H3 upgrade testing so I think it's low priority, but I agree.